### ISSUE TITLE
btrbk not waiting for autofs to mount

### ISSUE THREAD

**Issue by Massimo-B** [digint/btrbk#259]
Hi,

when using autofs targets, btrbk sometimes does not wait for getting the targets mounted and just skips the targets. Just the next run finds the target mounted and starts the send/receive.
Actually any ls on the target should block until autofs got it mounted, also 'btrfs filesystem usage' does so. I'm not sure if you have a timeout waiting for the target, that you can increase. Right now I just always do a 'btrfs filesystem usage' before starting btrbk. But then I need to add that also to the daily cronjobs doing btrbk in the background.

**Comment by Gatak** [comment]
Are you running this in cron? If so you could add checks in a script that runs btrbk.

**Comment by digint** [comment]
The problem is probably that btrbk outsmarts autofs by looking at `/proc/self/mountinfo` in order to know where to run `btrfs subvolume list` from (at least as of btrbk-0.27.0, earlier versions did that differently). I'm not sure how autofs works, but if there is no entry for your target mountpoint in mountinfo, then btrbk assumes that your target is mounted on "/" and lists the subvolumes from there.

If there is an entry in mountinfo, then maybe autofs simply does not intercept the system call used by `btrfs subvolume list <path>`, which is supposed to "list the subvolumes present in the filesystem \<path\>" (see btrfs-subvolume(8)). Now autofs either decides "nope, no mount needed, this command wants to list the filesystem at the currently mounted fs there", or it simply does not intercept the systemcall used for this. I suspect the latter. Or some systemd mis-magic.

The problem with autofs is, well, it never really works as you want it to (at least this is from my experience with it, some years ago). I don't know all the details of how it works, but there's a lot of magic ongoing between the kernel and userspace tools, with some added magic by systemd if you use it.

For the record: what kernel / autofs userspace / systemd versions do you use here?

**Comment by digint** [comment]
I found that btrbk bails out as soon as fs_type in `/proc/self/mountinfo` is not "btrfs", without even trying to access anything. This prevents autofs to mount.

Here's a patch which should solve the problem: 9615e64137b5f25d9fdf5fbebee9df31d677b8b9 (on branch "autofs-mount"). This calls `btrfs subvolume show` on the autofs mount point, which should trigger a mount.

@Massimo-B could you please try this:

```
cd /tmp/
wget https://raw.githubusercontent.com/digint/btrbk/autofs-mount/btrbk
chown +x btrbk
sudo ./btrbk dryrun -l debug
```
Does this automatically mount your autofs devices now and work as expected?

**Comment by Massimo-B** [comment]
This seems to work.

**Comment by digint** [comment]
thanks, merged "autofs-mount" branch to master: 17957a163a16836c3539cc1ef2830a2c5cc0e2d1

**Comment by digint** [comment]
implemented in btrbk-0.27.2, closing bug
